lib/katello/plugin.rb

```diff
@@ -668,6 +668,13 @@ def katello_template_setting_values(name)
         default: false,
         full_name: N_('Hide Reclaim Space Warning'),
         description: N_('If this is enabled, the Smart Proxy page will suppress the warning message about reclaiming space.')
+
+      setting 'deb_enable_structured_apt',
+        type: :boolean,
+        default: false,
+        full_name: N_('Enable structured APT for deb content'),
+        description: N_("APT repos in katello will have the same repo structure as the remote repos they are syncronized from. " \
+                        "This feature must be enabled/disabled by running 'foreman-rake katello:enable_structured_content_for_deb' in order to migrate existing deb content.")
```

Comment:
If it must be enabled via the rake task, why also have a Setting?

Reply:
To record somewhere if it is enabled, and also for users to be able to check somewhere if it is enabled. But if there is a better place for this (one where users cannot inadvertently change the setting without running the rake task), that would be better still. It may well be down to my lack of knowledge about suitable alternatives that I chose a setting.

Reply:
> if there is a better place for this (one where users cannot inadvertently change the setting without running the rake task), that would be better still.

The more I think about it the more I feel like a Setting isn't the right way to express this. Settings are for users to set, and they should take effect when you set them. So having a Setting here that does almost nothing just feels weird.

As I look at the rake task, it looks to me a bit like a database migration -- not exactly, since schema is not changing, but rather, data itself is being migrated. I think having a rake task to do this makes sense, but there should be a way other than a Setting to tell if it's been done or not. Ideally it would be something that could be some sort of computed value -- say, something like, `Katello::RootRepository.deb.any? { |root| root.content_id.present? }` If that's not reliable enough, perhaps we could add a boolean database field to RootRepository or something.

As for how to display it to the user, since it's on Repository maybe it would make sense to show something on the repository edit page. @quba42 @sjha4 thoughts?

Reply:
What you are saying definitely makes sense, I was never fully happy with the "setting that must not be altered via the UI"  aspect. However, having just done a round of rake task testing on a large installation, which ran into problems with "re-run robustness" (i.e. the rake task does not run to completion on the first attempt for some reason, and then needs to be re-run, but only for the remaining not yet migrated data), I would like to tweak your suggestion as follows:

- There is a computed value for each deb type repository that tells us if that repostiory is using "structured APT". (This does not require many changes, since `root.content_id.blank?` already answers the exact question we are asking here.)
- I keep the setting, but it is now interpreted as answering the question: should _newly created deb repos_ use "structured APT". (Users may freely toggle this setting in the UI).
- The rake task may be used to migrate any existing repositories that are in the other state from the setting, to match the current state of the setting.
- We could display the current state for each repo "structured APT enabled" vs. "structured APT disabled" on the repository page for each repo. (Not sure how important this is).

This approach decouples the question of "what state is a given deb repo in?" from the question of "would you like me to create new deb repos using the new feature?". This also means a partially migrated system (some repos have been migrated, but others are not) is no longer in a broken state (as is currently the case). This greatly lowers the stakes for the migration rake task. The worst thing the rake task can now possibly do is leave an individual repository in an inconsistent state, but it can't leave your entire DB in a globally inconsistent state that may be hard to recover from. It should be much easier to make the rake task robust for these "lower stakes".

Reply:
Ack to this modification. That makes much more sense to me. :+1:

Reply:
The suggestion is now implemented.